javascript/implementation/Role.ts

```diff
@@ -0,0 +1,124 @@
+
+    constructor(ginkInstance: GinkInstance, address: Muid, containerBuilder?: ContainerBuilder) {
+        super(ginkInstance, address, Behavior.ROLE);
+        if (this.address.timestamp < 0) {
+            ensure(address.offset == Behavior.ROLE);
+        } else {
+            ensure(containerBuilder.getBehavior() == Behavior.ROLE);
+        }
+    }
+
+    /**
+     * Includes a Muid or Container in the role.
+     * @param key either a container or a Muid to include
+     * @param change an optional bundler to put this change into
+     * @returns a promise that resolves to the Muid for the inclusion
+     */
+    async include(key: Muid|Container, change?: Bundler|string): Promise<Muid> {
+        return await this.addEntry(key, Container.INCLUSION, change);
+    }
+
+    /**
+     * Excludes a Muid or Container from the role.
+     * @param key either a Muid or container to exclude
+     * @param change an optional bundler to put this in
+     * @returns a promise that resolves to the Muid for the exclusion
+     */
+    async exclude(key: Muid|Container, change?: Bundler|string): Promise<Muid> {
+        return await this.addEntry(key, Container.DELETION, change);
+    }
+
+    /**
+     * The number of items in the role.
+     * @param asOf optional timestamp to look back to
+     * @returns a promise that resolves to the number of entries
+     */
+    async size(asOf?: AsOf): Promise<number> {
+        const entries = await this.ginkInstance.store.getKeyedEntries(this.address, asOf);
+        return entries.size;
+    }
+
+    /**
+     * Whether or not the given key is included in the role.
+     * @param key either a Muid or container to check if it is included
+     * @param asOf optional timestamp to look back to
+     * @returns a promise that resolves to a boolean stating whether the key is included
+     */
+    async contains(key: Muid|Container, asOf?: AsOf): Promise<boolean> {
+        if ("address" in key) {
+            key = key.address;
+        }
+        const entry = await this.ginkInstance.store.getEntryByKey(this.address, key, asOf);
+        return Boolean(entry);
+    }
+
+    /**
+     * Function to iterate over the contents of the role.
+     * @param asOf optional timestamp to look back to
+     * @returns an async iterator across everything in the role, with values returned as MuidTuples
+     */
+    get_member_ids(asOf?: AsOf): AsyncGenerator<MuidTuple|KeyType|[], void, unknown> {
+        const thisSet = this;
+        return (async function*(){
+            const entries = await thisSet.ginkInstance.store.getKeyedEntries(thisSet.address, asOf);
+            for (const [key, entry] of entries) {
+                yield entry.effectiveKey;
+            }
+        })();
+    }
+
+    /**
+     * Returns the content of the role as a set.
+     * @param asOf optional timestamp to look back to
+     * @returns a promise that resolves to a set of Muids
+     */
+    async toSet(asOf?: AsOf): Promise<Set<Muid>> {
```

Comment:
Rather than return a set of Muids it would probably be more useful to return an array of Containers.  Since you can't really test a set to see if it contains a Muid, really you'd only ever want to iterate over the things in the role.